### PR TITLE
UUID implementation - draft (#38)

### PR DESCRIPTION
**Pull request by pmarjan**
This PR drafts the implementation of the UUID for the ImportService module.

- Functionality is based on Magento\Framework\DataObject\IdentityGeneratorInterface::generateId()
- The source database table representation has one extra column of varbinary type called 'uuid'
- SourceRepository::getById() now loads by accepting a UUID string
- Data interfaces are changed accordingly
- The SourceUpload model validates the received UUID (if sent at all), and either passed it or generates new one

### CODE REVIEW

File: app/code/Magento/ImportService/Model/SourceUpload.php [hunk 1]
```diff
@@ -46,6 +64,10 @@ public function __construct(
     public function execute(SourceInterface $source)
     {
         try {
+            if (!$source->getUuid() || !$this->validator->validateUuid($source)) {
```
Comment (nuzil) [inline app/code/Magento/ImportService/Model/SourceUpload.php]: Not a good way to proceed with uuid is wrong. Please throw here an exception and do not proceed

File: app/code/Magento/ImportService/Model/SourceRepository.php [hunk 2]
```diff
@@ -84,12 +84,12 @@ public function save(SourceInterface $source)
      *
      * @throws NoSuchEntityException
      */
-    public function getById($id)
+    public function getById($uuid)
```
Comment (nuzil) [inline app/code/Magento/ImportService/Model/SourceRepository.php]: Please change it to getByUuid()

**Comment by nuzil** [comment]
Can you also merge 2.3-develop to this brunch and change all implementations to use UUID?

File: app/code/Magento/ImportService/Api/Data/SourceInterface.php [hunk 3]
```diff
@@ -14,7 +14,8 @@
  */
 interface SourceInterface extends ExtensibleDataInterface
 {
-    const ENTITY_ID = 'source_id';
+    const ENTITY_ID = 'entity_id';
```
[the remark this answers is not in the document]
Comment (pmarjan) [reply app/code/Magento/ImportService/Api/Data/SourceInterface.php]: This constant is references in _init method of ResourceModel/Source.php. Shall I still remove it?

File: app/code/Magento/ImportService/Api/Data/SourceUploadResponseInterface.php [hunk 4]
```diff
@@ -12,7 +12,7 @@
  */
 interface SourceUploadResponseInterface
 {
-    const SOURCE_ID = 'source_id';
+    const UUID = 'uuid';
```
[the remark this answers is not in the document]
Comment (pmarjan) [reply app/code/Magento/ImportService/Api/Data/SourceUploadResponseInterface.php]: Done.

File: app/code/Magento/ImportService/Model/Source/Validator.php [hunk 5]
```diff
@@ -5,8 +5,10 @@
  */
 namespace Magento\ImportService\Model\Source;
 
-use Magento\Framework\File\Mime\Proxy as Mime;
-use Magento\Framework\Filesystem\Driver\Http\Proxy as Http;
+use Magento\Framework\File\Mime;
+use Magento\Framework\File\Mime;
```
[the remark this answers is not in the document]
Comment (pmarjan) [reply app/code/Magento/ImportService/Model/Source/Validator.php]: Done.

File: app/code/Magento/ImportService/Model/SourceUpload.php (hunk 1 above)
Comment (pmarjan) [reply app/code/Magento/ImportService/Model/SourceUpload.php]: Identity generator is removed and exception is thrown.
The identity generator is added to Model\Import\Type\FileSourceType. Generation of uuid may occur in the save() method. Does it look OK like this?

File: app/code/Magento/ImportService/Model/SourceRepository.php (hunk 2 above)
Comment (pmarjan) [reply app/code/Magento/ImportService/Model/SourceRepository.php]: Done.

**Comment by pmarjan** [comment]
Yes, I did merge from remote upstream into local branch.

File: app/code/Magento/ImportService/Api/Data/SourceInterface.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (nuzil) [reply app/code/Magento/ImportService/Api/Data/SourceInterface.php]: No you are right. Leave it

File: app/code/Magento/ImportService/Model/Import/Type/FileSourceType.php [hunk 6]
```diff
@@ -77,17 +76,19 @@ private function generateFileName()
      */
     public function save(SourceInterface $source)
     {
+        /** @var string $uuid */
+        $uuid = $source->getUuid() ?: $this->identityGenerator->generateId();
+
         /** @var string $fileName */
-        $fileName = $this->generateFileName();
+        $fileName = $uuid;
```
Comment (nuzil) [inline app/code/Magento/ImportService/Model/Import/Type/FileSourceType.php]: Please set to import data correct file name, including ".csv" in this case

**Comment by nuzil** [comment]
@pmarjan can you please fix failed tests? (you can ignore static, its not related to your PR)

File: app/code/Magento/ImportService/Model/Import/Type/FileSourceType.php (hunk 6 above)
Comment (pmarjan) [reply app/code/Magento/ImportService/Model/Import/Type/FileSourceType.php]: Done

**Comment by pmarjan** [comment]
@nuzil 

This is a short summary of the failed tests

Jobs 100.6, 100.7, 100.8, 100.9 have this error in common:

`1. Magento\ImportService\Model\ImportRestartOperation::execute(int $uuid, string $serializedData) must be compatible with Magento\ImportService\Api\ImportRestartOperationInterface::execute(int $uuid, string $serializedData): void in /home/travis/build/magento/async-import/app/code/Magento/ImportService/Model/ImportRestartOperation.php on line 17`

This same error is probably why Jobs 100.12 and 100.13 occured: cannot move to production, since compilation cannot be completed.

I expect this is resolved by now with the latest commits/merges.

Job 100.8, 100.9 are showing

`4 Errors and 2 Failures coming from tests from Magento\Framework\MessageQueue\UseCase\
`
Job 100.10, 100.11

```
1 Failure

Magento\Setup\Declaration\WhitelistDeclarationTest::testConstraintsAndIndexesAreWhitelisted
New table elements that do not exist in the whitelist declaration:
{
    "import_service_source": {
        "constraint": [
            "IMPORT_SERVICE_SOURCE_UUID"
        ]
    }
}
Failed asserting that an array is empty.

```

Should I continue on any of these errors related to jobs 8,9,10 and 11?

**Comment by nuzil** [comment]
Hi @pmarjan 
Yes can you please fix last one failed test: 
```
1) Magento\Setup\Declaration\WhitelistDeclarationTest::testConstraintsAndIndexesAreWhitelisted
New table elements that do not exist in the whitelist declaration:
{
    "import_service_source": {
        "constraint": [
            "IMPORT_SERVICE_SOURCE_UUID"
        ]
    }
}
Failed asserting that an array is empty.
/home/travis/build/magento/async-import/dev/tests/integration/testsuite/Magento/Setup/Declaration/WhitelistDeclarationTest.php:87
```

And also please merge 2.3-develop into your branch, cause you have a conflicts with mainline

**Comment by pmarjan** [comment]
Hi @nuzil  

The WhitelistDeclarationTest error related tot the import_service_source should be gone now.
Locally, I got a new one from within the same test, related to "testmodule_default_hydrator_extension_attribute_entity", 

`New table elements that do not exist in the whitelist declaration:
{
    "testmodule_default_hydrator_extension_attribute_entity": {
        "constraint": [
            "FK_A59A4048B6782BFC544422BD59D9BE8B"
        ]
    }
}
Failed asserting that an array is empty.`

but I suppose that one is out of our scope here?

**Pull request merged**